sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuHashPartitioningBase.scala

```diff
@@ -60,34 +93,48 @@ abstract class GpuHashPartitioningBase(expressions: Seq[Expression], numPartitio
     }
   }
 
-  def partitionIdExpression: GpuExpression = GpuPmod(
-    GpuMurmur3Hash(expressions, GpuHashPartitioningBase.DEFAULT_HASH_SEED),
-    GpuLiteral(numPartitions))
+  def partitionIdExpression: GpuExpression = GpuPmod(hashFunc, GpuLiteral(numPartitions))
 }
 
-object GpuHashPartitioningBase {
+object GpuHashPartitioningBase extends Logging {
 
   val DEFAULT_HASH_SEED: Int = 42
 
-  def hashPartitionAndClose(batch: ColumnarBatch, keys: Seq[Expression], numPartitions: Int,
-      nvtxName: String, seed: Int = DEFAULT_HASH_SEED): PartitionedTable = {
-    val sb = SpillableColumnarBatch(batch, SpillPriorities.ACTIVE_ON_DECK_PRIORITY)
-    RmmRapidsRetryIterator.withRetryNoSplit(sb) { sb =>
-      withResource(sb.getColumnarBatch()) { cb =>
-        val parts = withResource(new NvtxRange(nvtxName, NvtxColor.CYAN)) { _ =>
-          withResource(GpuMurmur3Hash.compute(cb, keys, seed)) { hash =>
-            withResource(GpuScalar.from(numPartitions, IntegerType)) { partsLit =>
-              hash.pmod(partsLit, DType.INT32)
-            }
-          }
-        }
-        withResource(parts) { parts =>
-          withResource(GpuColumnVector.from(cb)) { table =>
-            table.partition(parts, numPartitions)
-          }
+  private[rapids] def toHashExpr(hashMode: HashMode, keys: Seq[Expression],
+      seed: Int = DEFAULT_HASH_SEED): GpuHashExpression = hashMode match {
+    case Murmur3Mode => GpuMurmur3Hash(keys, seed)
+    case HiveMode => GpuHiveHash(keys)
+    case _ => throw new UnsupportedOperationException(s"Unsupported hash mode: $hashMode")
+  }
+
+  private[rapids] def hashModeFromCpu(cpuHp: HashPartitioning, conf: RapidsConf): HashMode = {
+    var hashMode: HashMode = Murmur3Mode // default to murmur3
+    if (conf.isHashFuncPartitioningEnabled && conf.isIncompatEnabled) {
+      // One Spark distribution introduces a new field to define the hash function
+      // used by HashPartitioning. Since there is no shim for it, so here leverages
+      // Java reflection to access it.
+      try {
+        val hashModeMethod = cpuHp.getClass.getMethod("hashingFunctionClass")
+        hashMode = hashModeMethod.invoke(cpuHp) match {
+          case m if m == classOf[Murmur3Hash] => Murmur3Mode
+          case h if h == classOf[HiveHash] => HiveMode
+          case o => UnsupportedMode(o.asInstanceOf[Class[_]].getSimpleName)
         }
+        logInfo(s"Found hash function '$hashMode' from cpu hash partitioning.")
+      } catch {
+        case _: NoSuchMethodException => // not the customized spark distributions, noop
+          logInfo("No hash function field is found in cpu hash partitioning.")
```

Comment:
Nit: This is path for vanilla Spark, but the log seems like something is wrong.

Reply:
I will improve this in a follow-up.